### PR TITLE
Add NetworkPolicy

### PR DESCRIPTION
**Pull request by emiran-orange**
#### Overview

Add the ability to create a network policy with customizable egress

#### What this PR does / why we need it
Network policies are a best practice.
Specifically when deploying in a NS already containing network policies, it is required.

#### Special notes for your reviewer
N/A
#### Checklist

- [x] Change log updated in `Chart.yaml` (see the contributing guide for details)
- [x] Chart version bumped in `Chart.yaml` (see the contributing guide for details)
- [x] Documentation regenerated by running `make docs`

### CODE REVIEW

**Comment by sagikazarmark** [comment]
Network policy is extremely use case specific IMO. I haven't really seen open source projects bundling network policies with their deployment manifests, but I may be wrong.

Can you find some examples from the ecosystem?

**Comment by emiran-orange** [comment]
@sagikazarmark I wouldn't say that security is use case specific, having network policies enabled is a best practice.
Anyway, I haven't searched much but some of prometheus-community charts have it and almost all of bitnami charts have it.

File: charts/dex/templates/networkpolicy.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        - port: telemetry
+  egress:
+    - ports:
+        - port: 53
+          protocol: UDP
+        - port: 53
+          protocol: TCP
+    {{- if .Values.networkPolicy.egressRules }}
+    {{- toYaml .Values.networkPolicy.egressRules | nindent 4 }}
+    {{- end }}
```
Comment (mkilchhofer) [inline charts/dex/templates/networkpolicy.yaml:37]: I would not specify egress rules by default:

```suggestion
  {{- with .Values.networkPolicy.egressRules }}
  egress:
    {{- toYaml . | nindent 4 }}
  {{- end }}
```

File: charts/dex/templates/networkpolicy.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+apiVersion: networking.k8s.io/v1
+{{- end }}
+kind: NetworkPolicy
+metadata:
+  name: {{ include "dex.fullname" . }}
+  labels:
+    {{- include "dex.labels" . | nindent 4 }}
+spec:
+  policyTypes:
+    - Egress
```
Comment (mkilchhofer) [inline charts/dex/templates/networkpolicy.yaml:14]:
```suggestion
    {{- if .Values.networkPolicy.egressRules }}
    - Egress
    {{- end }}
```

File: charts/dex/values.yaml [hunk 3]
```diff
@@ -256,3 +256,16 @@ strategy: {}
   # rollingUpdate:
   #   maxUnavailable: 1
   # type: RollingUpdate
+
+# -- [Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) configuration (see [values.yaml](values.yaml) for details).
+# @default -- Disabled by default.
+networkPolicy:
+  enabled: false
+  egressRules: []
```
Comment (mkilchhofer) [inline charts/dex/values.yaml:264]:
```suggestion
  egressRules: []
  # Allow DNS egress traffic
  # - ports:
  #     - port: 53
  #       protocol: UDP
  #     - port: 53
  #       protocol: TCP
```

File: charts/dex/values.yaml [hunk 4]
```diff
@@ -256,3 +256,16 @@ strategy: {}
   # rollingUpdate:
   #   maxUnavailable: 1
   # type: RollingUpdate
+
+# -- [Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) configuration (see [values.yaml](values.yaml) for details).
+# @default -- Disabled by default.
+networkPolicy:
+  enabled: false
```
Comment (mkilchhofer) [inline charts/dex/values.yaml:263]:
```suggestion
networkPolicy:
  # -- Create [Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
  enabled: false
  # -- A list of network policy egress rules
```

**Review by mkilchhofer — CHANGES_REQUESTED** [review]

File: charts/dex/templates/networkpolicy.yaml (hunk 1 above)
Comment (emiran-orange) [reply charts/dex/templates/networkpolicy.yaml:37]: @mkilchhofer Thanks for the feedback.
Actually, as soon as any netpol is added to a NS, any other network flow is forbidden.
I thought of dns egress as as a failsafe but I'm fine with change things your way. I'm on it

Comment (mkilchhofer) [reply charts/dex/templates/networkpolicy.yaml]: Jup this is true :-) But most users have a global rule in place for DNS such as Calico's `GlobalNetworkPolicy` or Cilium's `CiliumClusterwideNetworkPolicy`.

Comment (emiran-orange) [reply charts/dex/templates/networkpolicy.yaml]: That should be it in fae3976

File: charts/dex/values.yaml [hunk 5]
```diff
@@ -256,3 +256,24 @@ strategy: {}
   # rollingUpdate:
   #   maxUnavailable: 1
   # type: RollingUpdate
+
+# -- [Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) configuration (see [values.yaml](values.yaml) for details).
+# @default -- Disabled by default.
```
Comment (mkilchhofer) [inline charts/dex/values.yaml:261]: Drop these lines and rerun helm-docs

Comment (emiran-orange) [reply charts/dex/values.yaml]: Arff sorry. Fixed in d08098b

**Review by mkilchhofer — DISMISSED** [review]
Lgtm 👍
@sagikazarmark please review again :)

Examples in CNCF's ecosystem:
- fluent-bit: https://github.com/fluent/helm-charts/blob/main/charts/fluent-bit/templates/networkpolicy.yaml
- Argo CD via kustomize: https://github.com/argoproj/argo-cd/blob/master/manifests/base/repo-server/argocd-repo-server-network-policy.yaml
- Argo CD via Helm: https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/values.yaml#L58
- giantswarm tools: https://github.com/giantswarm/starboard-exporter/blob/e42abdcf82c69945fbe3249dcd4d7abd7fc5e914/helm/starboard-exporter/templates/networkpolicy.yaml
- bitnami charts (matomo, wordpress, prestashop, drupal, mediawiki, ghost, etc.)
- grafana charts (grafana, promtail, loki-distributed)

**Review by mkilchhofer — DISMISSED** [review]

**Comment by emiran-orange** [comment]
@mkilchhofer I've just rebased my PR can you give your approval (again) ?

**Review by mkilchhofer — APPROVED** [review]
> @mkilchhofer I've just rebased my PR can you give your approval (again) ?

@emiran-orange
Sure, I can. But it doen't have any effect since I am not a maintainer of this project. I am just a random guy waiting for the NetworkPolicy to be implemented inside the chart ;-)

**Comment by emiran-orange** [comment]
@mkilchhofer Sorry about the misunderstanding!

**Comment by emiran-orange** [comment]
@sagikazarmark Have you seen https://github.com/dexidp/helm-charts/pull/74#pullrequestreview-1095560543 ? Are those examples enough for you ? If so, can you approve the PR ?

**Review by sagikazarmark — APPROVED** [review]
Thanks for the examples and the patience. Let's get this thing merged.

**Comment by mkilchhofer** [comment]
Maybe squash merge would be nice @sagikazarmark

**Pull request merged**